plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -436,19 +443,40 @@ else if (upperBound != null) {
         return Domain.create(ValueSet.ofRanges(statisticsRange), mayContainNulls);
     }
 
-    static boolean partitionMatchesConstraint(
-            Set<IcebergColumnHandle> identityPartitionColumns,
-            Supplier<Map<ColumnHandle, NullableValue>> partitionValues,
-            Constraint constraint)
+    private static class PartitionConstraintMatcher
     {
-        // We use Constraint just to pass functional predicate here from DistributedExecutionPlanner
-        verify(constraint.getSummary().isAll());
+        private final NonEvictableLoadingCache<Map<ColumnHandle, NullableValue>, Boolean> partitionConstraintResults;
+        private final Constraint constraint;
+
+        private PartitionConstraintMatcher(Constraint constraint)
+        {
```

Comment:
nit: require not null for constraint

Reply:
not needed, since constraint is invoked on within the ctor